### ISSUE TITLE
current uid is a stack

### ISSUE THREAD

**Issue by AndreasMadsen** [AndreasMadsen/dprof#10]
`MakeCallback` is sometimes called synchronously, thus `uid` can have multiple values in the same synchronous context.

**Comment by Fishrock123** [comment]
I think this should be fixed transitively via https://github.com/AndreasMadsen/dprof/issues/16

**Comment by AndreasMadsen** [comment]
I think they are quite unrelated. The fix is just:

```js
const currState = [root];

function asyncInit(uid, handle, provider, parentUid) {
  // get parent state
  const state = (parentUid === null ? currState[currState.length - 1] : stateMap.get(parentUid));

  // add new state node
  stateMap.set(uid, state.add(handle));
}

function asyncBefore(uid) {
  const state = stateMap.get(uid);

  state.before();
  currState.push(state);
}

function asyncAfter(uid) {
  const state = stateMap.get(uid);

  state.after();
  currState.pop();
}
```

but I haven't been able to create a test script for it.

**Comment by Fishrock123** [comment]
Sorry, wrong issue. I think this should be fixed by the actual async_hooks impl? aka https://github.com/nodejs/node/pull/8531

**Comment by AndreasMadsen** [comment]
Yes, it provided an API to get the contextual parent.

**Comment by AndreasMadsen** [comment]
Turns out that the basic `http` test does cover this after all. Fixed in a88b889551e2e56db7327afa4dab0245dde5e78e
